src/common/meta/src/key/datanode_table.rs

```diff
@@ -105,6 +106,96 @@ impl DatanodeTableManager {
+        distribution: RegionDistribution,
+    ) -> Result<()> {
+        let txns = distribution
+            .into_iter()
+            .map(|(datanode_id, regions)| {
+                let key = DatanodeTableKey::new(datanode_id, table_id);
+                let val = DatanodeTableValue::new(table_id, regions);
+
+                Ok(TxnOp::Put(key.as_raw_key(), val.try_as_raw_value()?))
+            })
+            .collect::<Result<Vec<_>>>()?;
+
+        txn.success.extend(txns);
+
+        Ok(())
+    }
+
+    /// Builds the update datanode table transactions. It only executes while the primary keys comparing successes.
+    pub(crate) fn build_update_txn(
+        &self,
+        txn: &mut TxnRequest,
+        table_id: TableId,
+        current_region_distribution: RegionDistribution,
+        new_region_distribution: RegionDistribution,
+    ) -> Result<()> {
+        let mut opts = Vec::new();
+
+        // Removes the old datanode table key value pairs
+        for current_datanode in current_region_distribution.keys() {
+            if !new_region_distribution.contains_key(current_datanode) {
+                let key = DatanodeTableKey::new(*current_datanode, table_id);
+                let raw_key = key.as_raw_key();
+                opts.push(TxnOp::Delete(raw_key))
+            }
+        }
+
+        for (datanode, regions) in new_region_distribution.into_iter() {
+            if let Some(current_region) = current_region_distribution.get(&datanode) {
+                // Updates if need.
+                if *current_region != regions {
+                    let key = DatanodeTableKey::new(datanode, table_id);
+                    let raw_key = key.as_raw_key();
+                    let val = DatanodeTableValue::new(table_id, regions).try_as_raw_value()?;
+                    opts.push(TxnOp::Put(raw_key, val));
+                }
+            } else {
+                // New datanodes
+                let key = DatanodeTableKey::new(datanode, table_id);
+                let raw_key = key.as_raw_key();
+                let val = DatanodeTableValue::new(table_id, regions).try_as_raw_value()?;
+                opts.push(TxnOp::Put(raw_key, val));
+            }
+        }
+
+        txn.success.extend(opts);
+        Ok(())
+    }
+
+    /// Builds the delete datanode table transactions. It only executes while the primary keys comparing successes.
+    pub fn build_delete_txn(
+        &self,
+        txn: &mut TxnRequest,
+        table_id: TableId,
+        distribution: RegionDistribution,
+    ) -> Result<()> {
+        let txns = distribution
+            .into_iter()
+            .map(|(datanode_id, regions)| {
+                let key = DatanodeTableKey::new(datanode_id, table_id);
+                let raw_key = key.as_raw_key();
+                let removed_key = to_removed_key(&String::from_utf8_lossy(&raw_key));
+                let val = DatanodeTableValue::new(table_id, regions);
+
+                Ok(TxnOp::Put(
```

Comment:
need to delete the original key

Reply:
Same reason, `DatanodeTable` is just the cache of the `TableRoute`. We can recover all `DatanodeTable` from the removed `TableRoute` key.